torch/storage.py

```diff
@@ -158,7 +158,10 @@ def _new_shared(cls, size):
 
 
 def _load_from_bytes(b):
-    return torch.load(io.BytesIO(b))
+    if torch.cuda.is_available():
```

Comment:
A change like this is going to be tricky to make. The context-manager in the issue is probably a safer solution, although it also requires careful commenting and documentation updates.

My concern with this is that whenever CUDA is unavailable this method starts loading tensors on the CPU, even if those tensors have another device type, like XLA. I don't think it's safe to do that.

Reply:
Hi,
first of all I agree that adding context awareness to the load procedure can be the optimal solution, but can be much more complicated to implement and verify.

I've trace the call to pickle load [this is the load implementation](https://github.com/python/cpython/blob/4ac923f2756f835f512339ee181348cc535ab07f/Lib/pickle.py#L1761) which is originated in my case in cloudpickle. I don't see a way we can pass custom arg there, I don't think it accepts *args, it also serves as a kind of middleware and does not really need to be aware if torch in my opinion.

here is another example of this issue in ray project (https://github.com/ray-project/ray/issues/9181).

the PR I suggested is safe, and even in the case you raised with a device of XLA, for what I can see in the documentation here https://pytorch.org/xla/release/1.7/index.html, it will still have a CPU device as a safe fall back.

this issue was originated when we train on cuda and try to load on a non-cuda device, so I try to think of a scenario when even in the presence of XLA device this solution wont work, as a cpu device will always be present and after we load our model to CPU we can we it to any device we want (XLA for example).

Reply:
It's true that the CPU device is always a fallback, but then why don't we always load and save tensors on the CPU? This PR would secretly create a law that only CUDA tensors can be restored to the device, because without CUDA loading would always secretly happen on the CPU. That's not a good UX.

Reply:
Yes I understand and agree, but also currently transition between devices will not be possible (through the usage of  _load_from_bytes).
what would you suggest to do to help implement this in a more user-friendly manner?
do you think it will be possible to check what device it was saved on (so that is the expected device to be loaded to) and if that is not available then fallback to CPU?

Reply:
Checking the device the tensor would like to restore to by default first and then warning that a CPU fallback is being used, for example, sounds reasonable.

Note that we do already have a good UX for this: torch.load, which supports map_location. This fix is only for users who are unable to use torch.load for some reason.